base/src/main/java/org/aya/generic/Shaped.java

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) 2020-2022 Yinsen (Tesla) Zhang.
+// Use of this source code is governed by the MIT license that can be found in the LICENSE.md file.
+package org.aya.generic;
+
+import org.aya.core.def.CtorDef;
+import org.aya.core.repr.AyaShape;
+import org.aya.core.term.CallTerm;
+import org.aya.core.term.ErrorTerm;
+import org.aya.core.term.Term;
+import org.aya.generic.util.InternalException;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.function.BiFunction;
+import java.util.function.Supplier;
+
+public interface Shaped<T> {
+  @NotNull AyaShape shape();
+  @NotNull Term type();
+  @NotNull T constructorForm();
+  <O> boolean sameEncoding(@NotNull Shaped<O> other);
+
+  interface Inductively<T> extends Shaped<T> {
+    @Override @NotNull Term type();
+    @NotNull T makeZero(@NotNull CtorDef zero);
+    @NotNull T makeSuc(@NotNull CtorDef suc, @NotNull T t);
+    @NotNull T destruct(int repr);
+    @NotNull T self();
+    int repr();
+
+    default @Override <O> boolean sameEncoding(@NotNull Shaped<O> other) {
+      if (shape() != other.shape()) return false;
+      if (!(other instanceof Inductively otherData)) return false;
+      var type = type();
+      var otherType = otherData.type();
+      return switch (type) {
+        case CallTerm.Data lhs && otherType instanceof CallTerm.Data rhs ->
+          lhs.ref().core == rhs.ref().core;
+        case CallTerm.Hole lhs && otherType instanceof CallTerm.Hole rhs ->
+          lhs.ref() == rhs.ref();
+        // TODO[literal]: different meta can have same solution
+        default -> false;
+      };
+    }
+
+    default <O> boolean sameValue(@NotNull Shaped<O> other) {
+      if (!sameEncoding(other)) return false;
+      var otherData = ((Inductively<O>) other);
+      return repr() == otherData.repr();
+    }
+
+    default @Override @NotNull T constructorForm() {
+      int repr = repr();
+      return with((zero, suc) -> {
+        if (repr == 0) return makeZero(zero);
+        return makeSuc(suc, destruct(repr - 1));
+      }, () -> {
+        // TODO[literal]: how to handle this?
+        throw new InternalException("trying to make constructor form without type solved");
+      });
+    }
+
+    default <R> R with(
```

Comment:
Do we have access to `TyckState`? If so, we should check solvedness using `TyckState`

Reply:
Now we will check solutions for terms, see https://github.com/aya-prover/aya-dev/pull/400/commits/6f576d6ca9912bd921f80bfabc0b9e89bcdebd1e